log/sequencer.go

```diff
@@ -401,7 +405,7 @@ func (s Sequencer) IntegrateBatch(ctx context.Context, tree *trillian.Tree, limi
 		}
 
 		// Collate node updates.
-		nodeMap, err := s.updateCompactTree(merkleTree, sequencedLeaves, label)
+		nodeMap, err := s.updateCompactTree(merkleTree, tree.TreeId, sequencedLeaves, label)
```

Comment:
_Ideally_ you probably only want this to happen once the `tx` has successfully committed?

It's probably fine because on a rollback the replayed root hash won't match the cached one so you'll end up having to init the cache from storage, but ...

Reply:
That was precisely my thinking. But it's probably still not a bad idea to make it safer. OTOH, it's not trivial: note that the returned `compact.Tree` is mutable, and we can't rollback these updates without some sort of "clone" function.

Reply:
This triggers another interesting observation: this code can be retried, as it's within a `RunInTransaction` function. If it is retried, the 1st retry will fail (and clear the cache) because of `compact.Tree` mutability and root hash mismatch, but the 2nd retry has a chance of succeeding again.

Reply:
Oh wait, ignore the last comment. The 1st retry will not fail, it will only be forced to reload the `compact.Tree` from storage.